s3_bucket: add encryption capabilities to the module

stefanhorning opened the pull request:
##### SUMMARY
Adds encryption capabilities to the s3_bucket module. Based on this PR #38538

##### ISSUE TYPE
- Feature Pull Request


##### COMPONENT NAME
s3_bucket

##### ADDITIONAL INFORMATION
Includes all change sfrom @julienvey made in PR #38538 and the changes from @wimnat made in his branch https://github.com/ansible/ansible/compare/devel...wimnat:s3_encryption?expand=1 and resolving of merge conflicts by me. To hopefully get this all into a mergable state.

julienvey committed:
s3_bucket: add encryption capabilities to the module

julienvey committed:
s3_bucket: updates after code review

wimnat committed:
Fix up for encrpytion based on review comments

stefanhorning committed:
Update version_added versions to 2.9

willthames commented on lib/ansible/modules/cloud/amazon/s3_bucket.py:
```diff
@@ -555,11 +631,13 @@ def main():
     )
 
-    module = AnsibleAWSModule(argument_spec=argument_spec)
+    module = AnsibleAWSModule(argument_spec=argument_spec, required_if=[['encryption', 'aws:kms', ['encryption_key_id']]])
```
better to split this onto two lines

willthames commented:
Looks good, I'll run the test suite

stefanhorning committed:
Keep the newer result management from devel branch.

stefanhorning commented:
I rebased this branch once more and resolved the last merging glitches now. @willthames also added the requested linebreak. So I guess this is ready to merge now.

stefanhorning committed:
Fix linting issue

stefanhorning commented:
I think s.b. needs to restart the build here (I didn't find a button on the shippable UI) as the failing test was working in the previous commit before I fixed the last remaining test for the linting, so technically the build is green here.

stefanhorning committed:
Just a commit to get build going again.

stefanhorning commented:
cc @wimnat @ryansb @s-hertel (previous reviewers)

stefanhorning commented:
ready_for_review

jillr commented on lib/ansible/modules/cloud/amazon/s3_bucket.py:
```diff
@@ -555,11 +636,15 @@ def main():
 
-    module = AnsibleAWSModule(argument_spec=argument_spec)
+    module = AnsibleAWSModule(
+        argument_spec=argument_spec, required_if=[['encryption', 'aws:kms', ['encryption_key_id']]]
```
```suggestion
        argument_spec=argument_spec, 
        required_if=[['encryption', 'aws:kms', ['encryption_key_id']]]
```
The line break should separate the args, see the style of other aws modules for examples.

jillr reviewed (changes requested):
tests good for me, thanks for this patch.

stefanhorning committed:
Add linebreak as suggested

stefanhorning commented:
@jillr Added the linebreak as suggested.

To everyone else: if there is no more linebreaks to add, please merge.

jillr reviewed (commented):
Once sanity tests pass, lgtm

stefanhorning commented:
Sanity tests are green now. But somehow integration tests started failing out of the blue. Was there recent change in the test environment in regards to AWS s3 permissions? If so they need to be rolled back to make the tests in here green again.

stefanhorning commented:
Rebased on latest devel to see if that fixes build issues.

stefanhorning committed:
Added s3:HeadBucket action to list, seems to get used in failing tests, eventhough this PR doesn't change anything in this regard

stefanhorning committed:
Put a pause into test to make it more robust.

stefanhorning committed:
Reverted change in storage-policy.json, use bucket names with dashes in integration test. Also added another pause to make tests run smoothly also in the aws docker envs

stefanhorning committed:
Testing

stefanhorning committed:
Bigger sleep for versioning apply, as it causes tests to fail

stefanhorning committed:
More sleeps in test

stefanhorning commented:
Ready to merge now @wimnat @ryansb @s-hertel @jillr @willthames
